lib/control/include/clingo/control/config.hh

```diff
@@ -0,0 +1,366 @@
+        //! Get the name of the nth subkey in this node.
+        [[nodiscard]] auto map_nth(KeyType index) const -> std::optional<std::string_view>;
+        //! Get the key for an array entry at the given index.
+        [[nodiscard]] auto array_at(Key key, KeyType index) const -> Key;
+        //! Get the value of this key.
+        auto get_value(std::optional<KeyType> index, std::string &value) const -> bool;
+        //! Set the value of this key.
+        void set_value(std::optional<KeyType> index, std::string_view value);
+        //! Get information about this key.
+        void info(Key key, int *n_children, int *array_info, ValueFlags *value_info) const;
+        //! Adds a subkey to this node.
+        void add_subkey(std::string_view name, Key parent, Key child);
+
+      private:
+        std::unique_ptr<Entry> entry_;
+        std::map<std::string, Key, std::less<>> subkeys_;
+        std::string description_;
+    };
+
+    //! Parses a configuration key name, extracting an optional index if present.
+    //!
+    //! Accepts key names in the form "key" or "key[index]". If an index is
+    //! present, it is parsed and returned (where an empty index represents an
+    //! invalid index). The input name is modified to remove any index portion.
+    //!
+    //! Throws if the key is malformed.
+    //!
+    //! @param name[inout] reference to the key name
+    //! @return an optional index value
+    static auto parse_name_(std::string_view &name) -> std::optional<IndexType>;
+
+    //! Resolves a clingo configuration key from a hierarchical path.
+    //!
+    //! Traverses the configuration tree starting from the given key and
+    //! following the path, which may contain dot-separated segments and
+    //! optional indices (e.g., "foo.bar[2].baz"). Returns a ClingoKey with the
+    //! resolved key id and index.
+    //!
+    //! The current implementation is limited to at most one index per key
+    //! segment.
+    //!
+    //! @param key the starting ClingoKey for resolution
+    //! @param path the dot-separated configuration path to resolve
+    //! @return the resolved ClingoKey
+    //! @throws std::runtime_error if the path is invalid
+    auto parse_path_(Key key, std::string_view path) const -> std::optional<Key>;
+    // Checks if this key is a clasp key.
+    //
+    // Returns true if the upper bit is not set.
+    [[nodiscard]] static auto is_clasp(KeyType key) -> bool;
+
+    // Checks if this key is a clingo key.
+    //
+    // Returns true if the upper bit is set and the key is not the invalid key.
+    [[nodiscard]] static auto is_clingo(KeyType key) -> std::optional<Key>;
+
+    // Get the string representation of the configuration tree under the given key.
+    void str_(Util::OutputBuffer &out, KeyType key, size_t first_indent, size_t indent) const;
+
+    Clasp::Cli::ClaspCliConfig *config_;
+    std::string mutable buf_;
```

Comment:
Is it guaranteed that objects of this class are never accessed from multiple threads?

Reply:
Imo it just makes no sense to provide a thread-safe config. I won't bother to support this.